crates/ruff/src/rules/pyupgrade/rules/use_pep585_annotation.rs

```diff
@@ -2,58 +2,71 @@ use rustpython_parser::ast::Expr;
-    name: String,
+    from: ModuleMember<'static>,
+    to: ModuleMember<'static>,
 }
 
 impl Violation for NonPEP585Annotation {
     const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
 
     #[derive_message_formats]
     fn message(&self) -> String {
-        let NonPEP585Annotation { name } = self;
-        format!(
-            "Use `{}` instead of `{}` for type annotations",
-            name.to_lowercase(),
-            name,
-        )
+        let NonPEP585Annotation { from, to } = self;
+        format!("Use `{to}` instead of `{from}` for type annotation")
     }
 
     fn autofix_title(&self) -> Option<String> {
-        let NonPEP585Annotation { name } = self;
-        Some(format!("Replace `{name}` with `{}`", name.to_lowercase()))
+        let NonPEP585Annotation { to, .. } = self;
+        Some(format!("Replace with `{to}`"))
     }
 }
 
 /// UP006
 pub(crate) fn use_pep585_annotation(checker: &mut Checker, expr: &Expr) {
-    if let Some(binding) = checker
-        .ctx
-        .resolve_call_path(expr)
-        .and_then(|call_path| call_path.last().copied())
-    {
-        let fixable = !checker.ctx.in_complex_string_type_definition();
-        let mut diagnostic = Diagnostic::new(
-            NonPEP585Annotation {
-                name: binding.to_string(),
-            },
-            expr.range(),
-        );
-        if fixable && checker.patch(diagnostic.kind.rule()) {
-            let binding = binding.to_lowercase();
-            if checker.ctx.is_builtin(&binding) {
-                #[allow(deprecated)]
-                diagnostic.set_fix(Fix::unspecified(Edit::range_replacement(
-                    binding,
+    let Some(SymbolReplacement { from, to }) = to_pep585_generic(expr, &checker.ctx) else {
+        return;
+    };
+
+    let fixable = !checker.ctx.in_complex_string_type_definition();
+    let mut diagnostic = Diagnostic::new(
+        NonPEP585Annotation {
+            from: from.clone(),
+            to: to.clone(),
+        },
+        expr.range(),
+    );
+    if fixable && checker.patch(diagnostic.kind.rule()) {
+        if to.is_builtin() {
+            // Built-in type, like `list`.
+            if checker.ctx.is_builtin(to.member()) {
+                diagnostic.set_fix(Fix::automatic(Edit::range_replacement(
+                    to.member().to_string(),
                     expr.range(),
                 )));
             }
+        } else {
+            // Imported type, like `collections.deque`.
+            diagnostic.try_set_fix(|| {
+                let (import_edit, binding) = get_or_import_symbol(
+                    to.module(),
+                    to.member(),
+                    expr.start(),
+                    &checker.ctx,
+                    &checker.importer,
+                    checker.locator,
+                )?;
+                let reference_edit = Edit::range_replacement(binding, expr.range());
+                Ok(Fix::suggested_edits(import_edit, [reference_edit]))
```

Comment:
What are the semantic changes that this fix can introduce? Does importing `collections` have any side effects? Or is it just because we feel less confident about fixes that patch up imports?

Reply:
It's mostly the latter -- I just want to improve the import-patching fixes before elevating them.